### ISSUE TITLE
OCC report:update IPC task did not complete successfully

### ISSUE THREAD

**Issue by lili-lilili** [open-power/hcode#5]
I'm testing powercapping on p10(opal).
If i set user cap limit to a small value(700w), occ always go to safe mode when htx test going.

OCC report the follow error log:
`task_core_data_control: pstate[0x79] update IPC task did not complete .`
It seems that PGPE in busy state.

Any suggestion?
I want to know what happened to PGPE.
But I don't know how to debug PGPE or how to obtain the PGPE log.

**Here is the occ logs:**
```
00000015.077730878|    1|ERR             |1494485872|E>cmdh_send_ambient_temp: ambient read failure status!
00000015.077733761|    1|INF             |1494516256|cmdh_send_ambient_temp new altitude 228m reference 1000m WOF ambient_adj_for_altitude[-2]
00000020.000729879|    1|INF             |1494496032|check_for_opal_updates: WOF state change 0x00->0x01
00000020.000733079|    1|INF             |1494607840|check_for_opal_updates: soft min Pcap = 0x02BC->0x02BC hard min Pcap = 0x0514->0x0514
00000020.000735612|    1|INF             |1494457776|check_for_opal_updates: max Pcap = 0x0700->0x0700 active Pcap = 0x0000->0x02D0
00000020.000738044|    1|INF             |1494562128|check_for_opal_updates: Pcap PSR = 100->100 Pcap source = 1->1
00000069.281019971|    0|INF             |1494539552|notify_host: notification of reason 0x01 has been sent
00000148.990219870|    0|IMP             |1494569776|First time throttling for reason[0x00004000] History[0x00004000] freq = 3884
00000155.837598419|    1|INF             |1494602320|notify_host: OCCMISC/core_ext_intr not clear yet (register=0x80010000, reason=0x01)
00000183.541631308|    1|INF             |1494602320|notify_host: OCCMISC/core_ext_intr not clear yet (register=0x80010000, reason=0x01)
00000800.712198926|    0|ERR             |1494546768|E>task_core_data_control: pstate[0x79] update IPC task did not complete successfully, idle?[0] rc[00000001]
00000800.712204439|    0|INF             |1494539392|createPgpeErrl: Creating error log in slot[0] max size[3616]
00000800.712207251|    0|IMP             |1494491584|createPgpeErrl: OISR0=0x00000020, OISR1=0x1000E550, OCCS2=0x02AE02BD
00000800.712210403|    0|INF             |1494605808|addPgpeDataToErrl: clipping pgpe trace size from 2104 to 992
00000800.712239836|    0|IMP             |1494435008|Activating reset required state.
00000800.712283859|    0|INF             |1494555120|Reporting error @ 0xFFFAB050 with size 2752
00000800.712286711|    0|INF             |1494569424|ModID: 0x00000406, RC: 0x000000D3, UserData1: 0x00000001, UserData2: 0x00000000
00000800.712370266|    1|IMP             |1494492720|SMGR: All States to Safe Transition Started
00000800.712373264|    1|INF             |1494525056|INTR: WOF disabled due to non-error condition. wof_disabled = 0x00000100
00000800.712375793|    1|ERR             |1494419824|E>INTR: WOF is being disabled. Reasoncode: 0x00000100
00000800.712378604|    1|INF             |1494559744|INTR: WOF disabled, setting frequency ranges
00000800.712382039|    1|INF             |1494410272|amec_set_freq_range: Mode[0x00] Fmin[2000/0x78] Fmax[2850/0x45]
```

**Comment by lili-lilili** [comment]
We found that sometimes cpu checkstop.
PEL log:
```
root@21B700016:~# peltool -i 0x500034E1
{
"Private Header": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Created by":               "0xE500",
    "Created at":               "01/30/2024 17:38:39",
    "Committed at":             "01/30/2024 17:38:39",
    "Creator Subsystem":        "BMC",
    "CSSVER":                   "",
    "Platform Log Id":          "0x500034E1",
    "Entry Id":                 "0x500034E1",
    "BMC Event Log Id":         "69190"
},
"User Header": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Log Committed by":         "0x2000",
    "Subsystem":                "Processor Unit (CPU)",
    "Event Scope":              "Entire Platform",
    "Event Severity":           "Unrecoverable Error",
    "Event Type":               "Not Applicable",
    "Action Flags": [
                                "Service Action Required",
                                "Report Externally",
                                "HMC Call Home"
    ],
    "Host Transmission":        "Not Sent",
    "HMC Transmission":         "Not Sent"
},
"Primary SRC": {
    "Section Version":          "1",
    "Sub-section type":         "1",
    "Created by":               "0xE500",
    "SRC Version":              "0x02",
    "SRC Format":               "0x55",
    "Virtual Progress SRC":     "False",
    "I5/OS Service Event Bit":  "False",
    "Hypervisor Dump Initiated":"False",
    "Backplane CCIN":           "232D",
    "Terminate FW Error":       "False",
    "Deconfigured":             "False",
    "Guarded":                  "True",
    "Error Details": {
        "Message":              "Error Signature: 0x20DA0020 0x00010001 0x50740A0E"
    },
    "Valid Word Count":         "0x09",
    "Reference Code":           "BD13E510",
    "Hex Word 2":               "00080055",
    "Hex Word 3":               "232D0010",
    "Hex Word 4":               "CC009544",
    "Hex Word 5":               "01000000",
    "Hex Word 6":               "20DA0020",
    "Hex Word 7":               "00010001",
    "Hex Word 8":               "50740A0E",
    "Hex Word 9":               "00000000",
    "Callout Section": {
        "Callout Count":        "1",
        "Callouts": [{
            "FRU Type":         "Normal Hardware FRU",
            "Priority":         "Medium Priority",
            "Location Code":    "U78DA.ND0.WZS01SL-P0-CPU0",
            "Part Number":      "F210110",
            "CCIN":             "AB42",
            "Serial Number":    "            "
        }]
    },
    "SRC Details": {
        "Primary Attention": "system checkstop",
        "Signature Description": {
            "Chip Desc": "node 0 proc 1 (P10 2.0)",
            "Signature": "EQ_L3_FIR(10)[14] L3 Dir read UE",
            "Attn Type": "checkstop"
        }
    }
},
"Extended User Header": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Created by":               "0x2000",
    "Reporting Machine Type":   "K1 Power",
    "Reporting Serial Number":  "21B700016",
    "FW Released Ver":          "",
    "FW SubSys Version":        "4.10.00",
    "Common Ref Time":          "00/00/0000 00:00:00",
    "Symptom Id Len":           "36",
    "Symptom Id":               "BD13E510_20DA0020_00010001_50740A0E"
},
"Failing MTMS": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Created by":               "0x2000",
    "Machine Type Model":       "K1 Power",
    "Serial Number":            "21B700016"
},
"User Data 0": {
    "Section Version": "1",
    "Sub-section type": "1",
    "Created by": "0x2000",
    "BMCLoad": "0.56 0.54 0.56",
    "BMCState": "Ready",
    "BMCUptime": "0y 0d 8h 4m 10s",
    "BootState": "OSRunning",
    "ChassisState": "On",
    "FW Version ID": "4.10.00",
    "HostState": "Running",
    "Process Name": "/usr/bin/openpower-hw-diags",
    "System IM": "60001000"
},
"User Data 1": {
    "Section Version": "1",
    "Sub-section type": "1",
    "Created by": "0x2000",
    "PEL_SUBSYSTEM": "0x13",
    "SRC6": "551157792",
    "SRC7": "65537",
    "SRC8": "1349782030",
    "_PID": "31953"
},
"User Data 2": {
    "Section Version": "1",
    "Sub-section type": "1",
    "Created by": "0x2000",
    "Data": [
        {
            "Deconfigured": false,
            "EntityPath": [
                38,
                1,
                0,
                2,
                0,
                5,
                1,
                35,
                2,
                83,
                1,
                7,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0
            ],
            "GuardType": "GARD_Unrecoverable",
            "Guarded": true,
            "LocationCode": "Ufcs-P0-CPU0",
            "Priority": "M"
        }
    ]
},
"User Data 3": {
    "Section Version": "1",
    "Sub-section type": "4",
    "Created by": "0xE500",
    "Hostboot Scratch Registers": {
        "0x0000283c": "0xaa811504",
        "0x000000004602f489": "0x0000000000000000"
    }
},
"User Data 4": {
    "Section Version": "1",
    "Sub-section type": "5",
    "Created by": "0xE500",
    "Scratch Register Error Signature": {
        "Chip ID": "0x004b0006",
        "Signature ID": "0x5993000a"
    }
},
"User Data 5": {
    "Section Version": "1",
    "Sub-section type": "3",
    "Created by": "0xE500",
    "Callout List FFDC": [
        {
            "Callout Type": "Hardware Callout",
            "Guard": true,
            "Priority": "medium",
            "Target": "physical:sys-0/node-0/proc-1/eq-2/fc-1/core-0"
        }
    ]
},
"User Data 6": {
    "Section Version": "1",
    "Sub-section type": "1",
    "Created by": "0xE500",
    "Signature List": [
        {
            "Chip Desc": "node 0 proc 0 (P10 2.0)",
            "Signature": "PB_EXT_FIR(0)[2] pb_x2_fir_err",
            "Attn Type": "checkstop"
        },
        {
            "Chip Desc": "node 0 proc 1 (P10 2.0)",
            "Signature": "EQ_L3_FIR(10)[14] L3 Dir read UE",
            "Attn Type": "checkstop"
        },
        {
            "Chip Desc": "node 0 proc 2 (P10 2.0)",
            "Signature": "PB_EXT_FIR(0)[4] pb_x4_fir_err",
            "Attn Type": "checkstop"
        },
        {
            "Chip Desc": "node 0 proc 3 (P10 2.0)",
            "Signature": "PB_EXT_FIR(0)[7] pb_x7_fir_err",
            "Attn Type": "checkstop"
        }
    ]
},
"User Data 7": {
    "Section Version": "1",
    "Sub-section type": "2",
    "Created by": "0xE500",
    "Register Dump": [
        "node 0 ocmb 6 (Explorer 2.0) *******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000",
        "node 0 proc 0 (P10 2.0) ************************************",
        "  GFIR_CS                   (0x570F001C) 1000 0000 0000 0000",
        "  CFIR_N1_CS                (0x03040000) 8000 0000 4000 0000",
        "  CFIR_N1_CS_MASK           (0x03040040) 2000 0000 0000 0000",
        "  PB_EXT_FIR                (0x030113AE) 2000 0000 0000 0000",
        "  PB_EXT_FIR_MASK           (0x030113B1) D400 0000 0000 0000",
        "node 0 ocmb 4 (Explorer 2.0) *******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000",
        "node 0 ocmb 0 (Explorer 2.0) *******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000",
        "node 0 ocmb 2 (Explorer 2.0) *******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000",
        "node 0 proc 1 (P10 2.0) ************************************",
        "  GFIR_CS                   (0x570F001C) 0000 0000 2000 0000",
        "  CFIR_EQ_CS                (0x22040000) 8001 0000 0000 0000",
        "  CFIR_EQ_CS_MASK           (0x22040040) 2000 0000 0000 0000",
        "  EQ_L3_FIR                 (0x22012600) 0003 1008 0000 0000",
        "  EQ_L3_FIR_MASK            (0x22012603) 4249 3E29 8000 0000",
        "  EQ_L3_FIR_ACT1            (0x22012607) 3DB4 4100 0000 0000",
        "node 0 ocmb 24 (Explorer 2.0) ******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000",
        "node 0 ocmb 26 (Explorer 2.0) ******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000",
        "node 0 ocmb 28 (Explorer 2.0) ******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000",
        "node 0 ocmb 30 (Explorer 2.0) ******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000",
        "node 0 proc 2 (P10 2.0) ************************************",
        "  GFIR_CS                   (0x570F001C) 1000 0000 0000 0000",
        "  CFIR_N1_CS                (0x03040000) 8000 0000 4000 0000",
        "  CFIR_N1_CS_MASK           (0x03040040) 2000 0000 0000 0000",
        "  PB_EXT_FIR                (0x030113AE) 0800 0000 0000 0000",
        "  PB_EXT_FIR_MASK           (0x030113B1) D400 0000 0000 0000",
        "node 0 ocmb 32 (Explorer 2.0) ******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000",
        "node 0 ocmb 34 (Explorer 2.0) ******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000",
        "node 0 ocmb 36 (Explorer 2.0) ******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000",
        "node 0 ocmb 38 (Explorer 2.0) ******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000",
        "node 0 proc 3 (P10 2.0) ************************************",
        "  GFIR_CS                   (0x570F001C) 1000 0000 0000 0000",
        "  CFIR_N1_CS                (0x03040000) 8000 0000 4000 0000",
        "  CFIR_N1_CS_MASK           (0x03040040) 2000 0000 0000 0000",
        "  PB_EXT_FIR                (0x030113AE) 0100 0000 0000 0000",
        "  PB_EXT_FIR_MASK           (0x030113B1) B400 0000 0000 0000",
        "node 0 ocmb 56 (Explorer 2.0) ******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000",
        "node 0 ocmb 58 (Explorer 2.0) ******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000",
        "node 0 ocmb 60 (Explorer 2.0) ******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000",
        "node 0 ocmb 62 (Explorer 2.0) ******************************",
        "  CHIPLET_OCMB_FIR_MASK     (0x08040002) 6627 FFE0 0000 0000"
    ]
}
}

```

**Comment by prasrang** [comment]
I can help to debug PGPE errors. Want some details
what is the Firmware build  used here and if you are able to do scom operation
then please run this 
getscom pu 64015 -all 

getscom pu 66015 -all

getscom pu 01060151 -all

getscom pu 6c0a6 -all

And also can you attach complete errorlog  details.

**Comment by lili-lilili** [comment]
> I can help to debug PGPE errors. Want some details what is the Firmware build used here and if you are able to do scom operation then please run this getscom pu 64015 -all
> 
> getscom pu 66015 -all
> 
> getscom pu 01060151 -all
> 
> getscom pu 6c0a6 -all


Thanks prasrang.
Our firmware is base on op-build-fw1030, and update occ to 706ae10(fw1050).

Here is the scom data:
root@21B700016:~# 
root@21B700016:~# getscom pu 64015 -all
pu	k0:n0:s0:p00       0x00031200FFF20014
pu	k0:n0:s0:p01       0x00031200FFF20014
pu	k0:n0:s0:p02       0x00031200FFF20014
pu	k0:n0:s0:p03       0x00011000FFF27944
/usr/bin/edbg getscom pu 64015 -all 

root@21B700016:~# getscom pu 66015 -all
pu	k0:n0:s0:p00       0x00021200FFF34014
pu	k0:n0:s0:p01       0x00021200FFF34014
pu	k0:n0:s0:p02       0x00021200FFF34014
pu	k0:n0:s0:p03       0x00021200FFF34014
/usr/bin/edbg getscom pu 66015 -all 

root@21B700016:~# getscom pu 01060151 -all
pu	k0:n0:s0:p00       0x0780078000000000
pu	k0:n0:s0:p01       0x0780078000000000
pu	k0:n0:s0:p02       0x0780078000000000
pu	k0:n0:s0:p03       0x0780078000000000
/usr/bin/edbg getscom pu 01060151 -all 

root@21B700016:~# getscom pu 6c0a6 -all
pu	k0:n0:s0:p00       0x02AE02BD00000000
pu	k0:n0:s0:p01       0x02AE02BD00000000
pu	k0:n0:s0:p02       0x02D102BD00000000
pu	k0:n0:s0:p03       0x02C702BD00000000
/usr/bin/edbg getscom pu 6c0a6 -all 

> 
> And also can you attach complete errorlog details.

What do you mean by a complete errorlog?

**Comment by prasrang** [comment]
From the above data, looks like system is not in failure state. Because OCC complex reset has happened.
So I can't tell what was wrong in  PGPE.

I need all the PEL logs from the system at the failure point.

**Comment by lili-lilili** [comment]
When occ entry safe mode, just one PEL log is found:
```
root@21B700016:~# 
root@21B700016:~# peltool -l
{
    "0x89000027": {
        "SRC":                  "BC8A1A56",
        "PLID":                 "0x89000027",
        "CreatorID":            "Hostboot",
        "Subsystem":            "HostBoot",
        "Commit Time":          "02/02/2024 08:20:12",
        "Sev":                  "Unrecoverable Error",
        "CompID":               "0x1A00"
    }
}
root@21B700016:~# peltool -i 0x89000027
{
"Private Header": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Created by":               "0x1A00",
    "Created at":               "00/00/0000 00:00:00",
    "Committed at":             "02/02/2024 08:20:12",
    "Creator Subsystem":        "Hostboot",
    "CSSVER":                   "",
    "Platform Log Id":          "0x89000027",
    "Entry Id":                 "0x89000027",
    "BMC Event Log Id":         "69648"
},
"User Header": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Log Committed by":         "0x1A00",
    "Subsystem":                "HostBoot",
    "Event Scope":              "Entire Platform",
    "Event Severity":           "Unrecoverable Error",
    "Event Type":               "Not Applicable",
    "Action Flags": [
                                "Service Action Required",
                                "Report Externally",
                                "HMC Call Home"
    ],
    "Host Transmission":        "Not Sent",
    "HMC Transmission":         "Not Sent"
},
"Primary SRC": {
    "Section Version":          "1",
    "Sub-section type":         "1",
    "Created by":               "0x0",
    "SRC Version":              "0x02",
    "SRC Format":               "0xE0",
    "Virtual Progress SRC":     "False",
    "I5/OS Service Event Bit":  "False",
    "Hypervisor Dump Initiated":"False",
    "Valid Word Count":         "0x09",
    "Reference Code":           "BC8A1A56",
    "Hex Word 2":               "000000E0",
    "Hex Word 3":               "00003400",
    "Hex Word 4":               "00000000",
    "Hex Word 5":               "00200000",
    "Hex Word 6":               "00000000",
    "Hex Word 7":               "00000000",
    "Hex Word 8":               "00000000",
    "Hex Word 9":               "00000000",
    "Callout Section": {
        "Callout Count":        "1",
        "Callouts": [{
            "FRU Type":         "Maintenance Procedure Required",
            "Priority":         "Mandatory, replace all with this type as a unit",
            "Procedure":        "HB00055"
        }]
    }
},
"Extended User Header": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Created by":               "0x1A00",
    "Reporting Machine Type":   "FP5280G3",
    "Reporting Serial Number":  "B700016",
    "FW Released Ver":          "NL1030_026",
    "FW SubSys Version":        "1.07.00",
    "Common Ref Time":          "00/00/0000 00:00:00",
    "Symptom Id Len":           "76",
    "Symptom Id":               "BC8A1A56_000000E000003400000000000020000000000000000000000000000000000000"
},
"Extended User Data": {
    "Section Version": "1",
    "Sub-section type": "1",
    "Created by": "0x2000",
    "BMCState": "Ready",
    "BootState": "OSRunning",
    "ChassisState": "On",
    "FW Version ID": "4.10.00",
    "HostState": "Running",
    "System IM": "60001000"
},
"Failing MTMS": {
    "Section Version":          "1",
    "Sub-section type":         "0",
    "Created by":               "0x1A00",
    "Machine Type Model":       "FP5280G3",
    "Serial Number":            "B700016"
},
"User Data 0": {
    "Section Version": "1",
    "Sub-section type": "6",
    "Created by": "0x0100",
    "Callout Type": "Procedure Callout",
    "Procedure": "EPUB_PRC_HB_CODE",
    "Priority": "SRCI_PRIORITY_HIGH",
    "Flag": ""
},
"User Data 1": {
    "Section Version": "1",
    "Sub-section type": "1",
    "Created by": "0x0100",
    "String Data 1": "Hostboot Build ID: hostboot-p10-a57745-main-pd180a4f/hbirt.bin",
    "String Data 2": "FW Long ID: ba1b4554",
    "String Data 3": "FW Short ID: 1.07.00"
},
"User Data 2": {
    "Section Version": "1",
    "Sub-section type": "21",
    "Created by": "0x3100",
    "Decode file used": "/var/lib/phosphor-software-manager/hostfw/running/81e00685.lid",
    "Data": [
        "-------------------------------------------------------------------------------",
        "TRACEBUFFER: tracBINARY",
        "-------------------------------------------------------------------------------",
        "   Sec      Usec    PID  Comp          Line Entry   Data",
        "-------------------------------------------------------------------------------",
        "00000124.597355338|    0|RUNTIME         |1158|E>createPMCLoadStartCallback: host_callback interface not linked"
    ]
}
}
root@21B700016:~# 
```

**Comment by dcrowell77** [comment]
That error would indicate that the host (opal I assume) didn't register the "host_callback" interface for HBRT to use.  That was added in P10 so I'm not surprised they don't have it.  However, the OCC should have already been started during the IPL instead in Opal mode so this error shouldn't have any detrimental effect on the PM complex.  I would not be at all surprised if there were a few more spots up in HBRT that need opal/phyp checks.

**Comment by dcrowell77** [comment]
The HBRT (opal-prd) traces should include any deliberate attempts to reset the OCC/PM Complex.  The request typically would come from the HTMGT code due to some kind of occ attention.

You should probably confirm with the ipl hostboot traces that the pm complex is actually getting started in the IPL.  That code path has not been tested in P10 by IBM.

**Comment by lili-lilili** [comment]
> That error would indicate that the host (opal I assume) didn't register the "host_callback" interface for HBRT to use. That was added in P10 so I'm not surprised they don't have it. However, the OCC should have already been started during the IPL instead in Opal mode so this error shouldn't have any detrimental effect on the PM complex. I would not be at all surprised if there were a few more spots up in HBRT that need opal/phyp checks.

Thanks Dan.
I agree with what you said.
Our problem is that: 
machine has boot to OS success, and the OCC running normally.
But when we enable powercapping  and set a smaller limit value, then the OCC report error "task_core_data_control: pstate[0x79] update IPC task did not complete". I think that why occ go to safe mode.

I think we need to figure out what happened to PGPE first.

**Comment by prasrang** [comment]
what is the smaller limit value you are trying to set ? 

I think I remember there was a similar issue , that got reported from linux team here. And I had provided one fix, but that is in FW1060. If you have access to our HCODE and able to build , then I can mention the changes and file name. So that you can try once.

**Comment by lili-lilili** [comment]
> what is the smaller limit value you are trying to set ?
> 
Run htx test on the machine, the full power is about 1000w.
If i set a powercapping vaule less than 700w, there will be a probability of problems occurring(occ entry safe mode or cpu checkstop).
> I think I remember there was a similar issue , that got reported from linux team here. And I had provided one fix, but that is in FW1060. If you have access to our HCODE and able to build , then I can mention the changes and file name. So that you can try once.

Is IBM private repository? I can only access to this one:https://github.com/open-power/hcode?

**Comment by dcrowell77** [comment]
All of the code is out in https://github.com/open-power/hcode.  You likely just need to pull from master-p10 vs one of the branches to see a recent change.

**Comment by prasrang** [comment]
I think there will be some limit value where you can move up or move down. I believe the VPD data ( #V data which will be used by PGPE) will be characterized  based on the some value (threshold value). what if we go down below that threshold value, then PGPE should put the frequency to safe mode value.    

I think in this area we had issue in one of the linux system, where OCC was requesting the  pstate value(frequency) where there was no data for that in the VPD , that we refer for the corresponding voltage value and other parameters.

May be I want to try this fix , I believe that fix is not there in 1050, but it's in 1060. Do you guys 1060 drivers as well or only 1050?

**Comment by lili-lilili** [comment]
> Do you guys 1060 drivers as well or only 1050

I haven't try hcode-1060.
Let's me confirm the fix has push to master-p10 branch of https://github.com/open-power/hcode?

**Comment by prasrang** [comment]
I do see master-p10 hcode has that fix.  May be if you could try that, it will be good.

**Comment by lili-lilili** [comment]
I'm sorry for replying so late because I just finished my vacation.
I have test hcode master-p10 just now.
Unfortunately, it cannot solve the problem.

The error log remains the same as before.

**Comment by prasrang** [comment]
Could you attach the complete log

**Comment by lili-lilili** [comment]
> Could you attach the complete log

What logs do you need？

OCC log:
```
-------------------------------------------------------------------------------
TRACEBUFFER: Mixed buffer
-------------------------------------------------------------------------------
 Sec    Usec      PID Comp             Line Entry Data
-------------------------------------------------------------------------------
00000000.000123975|    1|IMP             |323710912|Proc ChipId (0)  NodeId (0)
00000000.000276048|    1|IMP             |323720192|Sensor Initialization Complete
00000000.000355070|    1|IMP             |323645808|MAIN: read_oppb_params: PGPE Frequency data: min[1999000], max[4000000], step[16667] kHz, pState min[0x78]
00000000.000357136|    1|IMP             |323774896|MAIN: populate_sys_mode_freq_table: Clipping freq to system max ceiling[3900000KHz] < Pstate0[4000000KHz]
00000000.000359477|    1|IMP             |323772096|MAIN: populate_sys_mode_freq_table: clipping chip Fmax[4000MHz] to system max[3900MHz]
00000000.000361477|    1|ERR             |323763088|E>MAIN: populate_sys_mode_freq_table: chip min[2000MHz] is higher than sys min[1999MHz]
00000000.000363870|    1|IMP             |323712896|MAIN: populate_sys_mode_freq_table: freq points min[2000MHz], WOF Base[2850MHz], UT[3900MHz], max[3900MHz]
00000000.000366023|    1|IMP             |323787152|MAIN: populate_sys_mode_freq_table: Mode freq disabled[2600MHz], Psav[2000MHz], Perf[3900MHz], Fmax[3900MHz]
00000000.000373253|    1|IMP             |323626464|MAIN: read_oppb_params:  pstate_max_throttle[0xB8]/1999000kHz(64 steps)  Fmin Pstate[0x78]
00000000.000383672|    1|IMP             |323757088|MAIN: PGPE HEADER magic number[0x504750455F312E30]
00000000.000385580|    1|IMP             |323777824|MAIN: read_pgpe_header: PGPE Beacon Address[0xFFF33804]
00000000.000387521|    1|IMP             |323798336|MAIN: read_pgpe_header: Valid PGPE Shared SRAM address[0xF7F33800] and Magic Number[0x4F505332]
00000000.000389633|    1|IMP             |323761040|MAIN: read_pgpe_header: Valid HCODE Elog version[0x454C5443], HCODE Elog Table [0xF7F33868] (11 slots)
00000000.000391754|    1|IMP             |323722160|MAIN: read_pgpe_header: Produced WOF values SRAM address OCC[0xF7F33820] PGPE[0xF7F33830] XGPE[0xF7F33828]
00000000.000393953|    1|IMP             |323620816|MAIN: read_pgpe_header: IDDQ Activity SRAM addr[0xF7F338C0] depth[0x08] bit shift[3]
00000000.000395804|    1|IMP             |323741376|MAIN: read_pgpe_header: Pstate table SRAM Address[0xFFF33940]
00000000.000410192|    1|IMP             |323824336|MAIN: read_wof_header: OPPB has VCS disabled using WOF header TDP index 0x02
00000000.000412499|    1|INF             |323672896|MAIN: Initializing watchdog timers.
00000000.000414392|    1|IMP             |323761776|Initializing ppc405 watchdog. period=3, reset_ctrl=0
00000000.000437408|    1|INF             |323778896|MAIN: timer created and scheduled
00000000.000441603|    1|INF             |323751696|CMDH: Command Handler Thread Started ...
00000000.000445139|    1|INF             |323615200|CMDH: Cmd_Hndl_thread_routine: OCB_OCBSHCS1=0x4FE00001
00000000.000447104|    1|INF             |323709792|cmdh_thread_wait_for_wakeup: OCB_OCBSHCS1=0x4FE00001
00000000.000463612|    1|INF             |323817296|DCOM: DCOM Thread Started
00000000.000466328|    1|INF             |323788256|check_for_opal_updates: OCC state change 0x00->0x01
00000000.000468236|    1|INF             |323791600|check_for_opal_updates: Unable to enable WOF due to WOF disable[0x00000400]
00000000.000470499|    1|INF             |323819488|check_for_opal_updates: soft min Pcap = 0x0000->0x0000 hard min Pcap = 0x0000->0x0000
00000000.000472734|    1|INF             |323669424|check_for_opal_updates: max Pcap = 0x0000->0x0000 active Pcap = 0x0000->0x0000
00000000.000474876|    1|INF             |323773776|check_for_opal_updates: Pcap PSR = 0->100 Pcap source = 0->0
00000010.000465422|    1|INF             |323751200|notify_host: notification of reason 0x01 has been sent
00000010.264529921|    0|INF             |323709472|notifyFspDoorbellReceived: OCB_OCBSHCS1=0x4FE04041
00000010.264565432|    1|INF             |323767728|cmdh_fsp_cmd_hndler: HTMGT command(cmd: 0x00, seq: 0x01)
00000010.264597662|    1|INF             |323814128|cmdh_fsp_cmd_hndler: HTMGT response(status: 0x00, dataLen: 0x01CE)
00000010.764490784|    1|IMP             |323812144|CMDH: Data Config Packet Received Type: 0x0f
00000010.764494158|    1|INF             |323728864|createErrl: Creating error log in slot[0] max size[3616]
00000010.764537745|    1|INF             |323766768|Reporting error @ 0xFFFAB050 with size 1752
00000010.764540394|    1|INF             |323781072|ModID: 0x00001103, RC: 0x000000DF, UserData1: 0x00000400, UserData2: 0x00000000
00000010.764543063|    1|INF             |323813968|notify_host: OCCMISC/core_ext_intr not clear yet (register=0x9C000000, reason=0x04)
00000010.764547191|    1|IMP             |323773440|CMDH: Got valid System Config data via TMGT for system type: 0x81
00000010.973229223|    1|INF             |323724592|deleting error @0xFFFAB050 at slot [0]
00000011.076618592|    1|IMP             |323812144|CMDH: Data Config Packet Received Type: 0x04
00000011.076622252|    1|INF             |323645280|CMDH: apss_store_ipmi_sensor_id: No Sensor ID for channel 0.
00000011.076624954|    1|INF             |323645280|CMDH: apss_store_ipmi_sensor_id: No Sensor ID for channel 1.
00000011.076627576|    1|INF             |323645280|CMDH: apss_store_ipmi_sensor_id: No Sensor ID for channel 2.
00000011.076630222|    1|INF             |323645280|CMDH: apss_store_ipmi_sensor_id: No Sensor ID for channel 3.
00000011.076632920|    1|INF             |323645280|CMDH: apss_store_ipmi_sensor_id: No Sensor ID for channel 4.
00000011.076635539|    1|INF             |323645280|CMDH: apss_store_ipmi_sensor_id: No Sensor ID for channel 5.
00000011.076638556|    1|INF             |323645280|CMDH: apss_store_ipmi_sensor_id: No Sensor ID for channel 7.
00000011.076642055|    1|INF             |323645280|CMDH: apss_store_ipmi_sensor_id: No Sensor ID for channel 11.
00000011.076644759|    1|INF             |323645280|CMDH: apss_store_ipmi_sensor_id: No Sensor ID for channel 12.
00000011.076647381|    1|INF             |323645280|CMDH: apss_store_ipmi_sensor_id: No Sensor ID for channel 13.
00000011.076650286|    1|INF             |323645280|CMDH: apss_store_ipmi_sensor_id: No Sensor ID for channel 15.
00000011.076654567|    1|IMP             |323745056|CMDH: data_store_apss_config: Enabling Vdd power cap[0625W] drop[0087 mhz] raise[0017 mhz]
00000011.076657061|    1|IMP             |323610112|CMDH: data_store_apss_config: Alg wait times decrease[0005 ticks] increase[0001 ticks] Proportional[00]
00000011.076659297|    1|IMP             |323731216|CMDH: Got valid APSS Config data via TMGT; Pwr reading type = 0
00000011.283215569|    1|IMP             |323812144|CMDH: Data Config Packet Received Type: 0x03
00000011.283218412|    1|INF             |323760672|initialize_apss: Creating/Scheduling request for GPE_apss_initialize_gpio
00000011.283309496|    1|INF             |323765056|initialize_apss: GPE_apss_initialize_gpio completed w/rc=0x00000082
00000011.283311980|    1|INF             |323734688|initialize_apss: Creating/Scheduling request for GPE_apss_set_mode
00000011.283335431|    1|INF             |323630784|initialize_apss: GPE_apss_set_mode completed
00000011.283337713|    1|INF             |323716128|initialize_apss: Creating request G_meas_start_request.
00000011.283340240|    1|INF             |323763232|initialize_apss: Creating request G_meas_cont_request.
00000011.283342748|    1|INF             |323749056|initialize_apss: Creating request G_meas_complete_request.
00000011.283345260|    1|INF             |323799648|initialize_apss: Creating request G_init_gpio_request for active state.
00000011.283347729|    1|INF             |323768240|initialize_apss: Creating request G_init_mode_request for active state.
00000011.283350209|    1|INF             |323624272|initialize_apss: Creating request G_reset_request.
00000011.283353493|    1|IMP             |323827104|CMDH: data_store_role: OCC Role set to Master via TMGT
00000011.284421804|    0|IMP             |323797328|Slave OCC[0] Received first doorbell from Master OCC[0]
00000011.284989191|    0|IMP             |323629632|Updated OCCs Present -- OldMask: 0x01, NewMask: 0x0f
00000011.490316846|    1|IMP             |323812144|CMDH: Data Config Packet Received Type: 0x05
00000011.490324236|    1|INF             |323797632|CMDH: data_store_mem_cfg: Setting mem_type from 0xA0 to I2C 0xC0
00000011.490326748|    1|IMP             |323706304|CMDH: data_store_mem_cfg: Got valid mem cfg packet. type=0xC0, #mem bufs=4, #dimm=4
00000011.490329288|    1|IMP             |323703728|CMDH: 1 OCMB will be read every 16ms Each OCMB read every 256ms (3906 reads/1000s)
00000011.490331715|    1|IMP             |323623936|CMDH: Memory monitoring is allowed (mem config data sets = 8, ips_mem_pwr_ctl = 255, default_mem_pwr_ctl = 255)
00000011.696919864|    1|IMP             |323812144|CMDH: Data Config Packet Received Type: 0x07
00000011.696922832|    1|IMP             |323735024|CMDH: data store pcap: Got valid PCAP Config data via TMGT. Count:1, Data Cfg mask[22c]
00000011.697621806|    0|IMP             |323629008|amec_data_write: PCAP Config data: pcap[2240]: data_mask[23c]
00000011.697659139|    1|INF             |323707680|check_for_opal_updates: WOF state change 0xE0->0x00
00000011.697662042|    1|INF             |323819488|check_for_opal_updates: soft min Pcap = 0x0000->0x01F4 hard min Pcap = 0x0000->0x0640
00000011.697664718|    1|INF             |323669424|check_for_opal_updates: max Pcap = 0x0000->0x08C0 active Pcap = 0x0000->0x0000
00000011.697667313|    1|INF             |323773776|check_for_opal_updates: Pcap PSR = 100->100 Pcap source = 0->0
00000011.904914600|    1|IMP             |323812144|CMDH: Data Config Packet Received Type: 0x12
00000011.904918931|    1|IMP             |323820336|CMDH: data_store_mem_throt: Got valid mem throt packet. configured_mba_bitmap=0x0055
00000012.112091351|    1|IMP             |323812144|CMDH: Data Config Packet Received Type: 0x13
00000012.112094760|    1|INF             |323736208|CMDH: data_store_thrm_thresholds: FRU_type[0xF0] DVFS Delta[0] Error Delta[10]
00000012.112097232|    1|INF             |323741072|CMDH: data_store_thrm_thresholds: FRU_type[0xF0] VPD Temp[85] DVFS[85] Error[95]
00000012.112100693|    1|INF             |323736208|CMDH: data_store_thrm_thresholds: FRU_type[0xF9] DVFS Delta[0] Error Delta[10]
00000012.112103144|    1|INF             |323741072|CMDH: data_store_thrm_thresholds: FRU_type[0xF9] VPD Temp[85] DVFS[85] Error[95]
00000012.112105504|    1|IMP             |323781248|CMDH: data_store_thrm_thresholds: Got valid Thermal Control Threshold data packet
00000012.112108099|    1|INF             |323811856|AMEC_data_write_thrm_thresholds: Processor setpoints - DVFS: 85, Error: 95
00000012.112110408|    1|INF             |323820944|AMEC_data_write_thrm_thresholds: membuf setpoints - DVFS: 84, Error: 87
00000012.112112720|    1|INF             |323631632|AMEC_data_write_thrm_thresholds: DIMM setpoints - DVFS: 77, Error: 80
00000012.112115032|    1|INF             |323728720|AMEC_data_write_thrm_thresholds: MC+DIMM setpoints - DVFS: 65, Error: 65
00000012.112117338|    1|INF             |323648448|AMEC_data_write_thrm_thresholds: PMIC setpoints - DVFS: 77, Error: 80
00000012.112119644|    1|INF             |323653536|AMEC_data_write_thrm_thresholds: External MC setpoints - DVFS: 65, Error: 65
00000012.112121953|    1|INF             |323698800|AMEC_data_write_thrm_thresholds: VRM Vdd setpoints - DVFS: 85, Error: 95
00000012.112124263|    1|INF             |323823888|AMEC_data_write_thrm_thresholds: Processor IO setpoints - DVFS: 85, Error: 95
00000012.318688074|    1|IMP             |323812144|CMDH: Data Config Packet Received Type: 0x14
00000012.318691107|    1|INF             |323819600|avsbus_init: Vdd=Y
00000012.318693544|    1|INF             |323621648|avsbus_init: reg[OCB_O2SCTRL10] = 0xD06C4000
00000012.318695902|    1|INF             |323712704|avsbus_resync: Send idle frame (bus 0)
00000012.318712469|    1|IMP             |323790928|CMDH: data_store_avsbus_config: Got valid AVS Bus data packet
00000012.318834929|    0|IMP             |323701808|Setting Active Bit for this OCC
00000012.525312871|    1|IMP             |323812144|CMDH: Data Config Packet Received Type: 0x15
00000012.525315831|    1|IMP             |323675312|CMDH: data_store_gpu: I2C engine = 0x01 I2C bus voltage = 0 deci volts
00000012.525318316|    1|IMP             |323778752|CMDH: data_store_gpu: Got valid GPU data packet
00000019.032437655|    1|INF             |323699392|User selected power limit = 700 set from source 1
00000019.033456298|    0|IMP             |323629008|amec_data_write: PCAP Config data: pcap[700]: data_mask[efc]
00000019.033493537|    1|INF             |323813968|notify_host: OCCMISC/core_ext_intr not clear yet (register=0x9C000000, reason=0x01)
00000019.962486375|    1|INF             |323816512|CMDH: SMS Mode=0, State=3, Mode Parm[0x0000]
00000019.963328730|    1|IMP             |323623264|SMGR: Standby to Active Transition Started
00000019.963331328|    1|IMP             |323752352|SMGR: Standby to Observation Transition Started
00000019.963333838|    1|INF             |323696912|ocmb_init: Initializing Memory Data Controller
00000019.963337084|    1|INF             |323778144|membuf_configuration_create: Scheduling request for IPC_ST_MEMBUF_INIT_FUNCID
00000019.963476913|    1|INF             |323620464|membuf_configuration_create: GPE_membuf_configuration_create w/rc=0x00000082
00000019.963480983|    1|IMP             |323702752|ocmb_init: G_present_membufs = 0x0000aa00
00000019.963483632|    1|IMP             |323670912|bitmap of enabled dimm temperature sensors: 0x0000000000000000 0000000000000000
00000019.963517344|    1|IMP             |323747040|SMGR: Standby to Observation Transition Completed
00000019.963519952|    1|IMP             |323743760|SMGR: Observation to Active Transition Started
00000019.963523084|    1|IMP             |323711408|pgpe_clip_update: Scheduling clip update: min[0xB8], max[0x45]
00000019.963538581|    1|IMP             |323715984|SMGR_obs_to_active: Successfully set Pstate clip max[0x45]
00000019.963541392|    1|IMP             |323654000|pgpe_start_suspend: scheduling enable of pstates (owner=0x01)
00000019.963841863|    1|IMP             |323649296|set_nominal_pstate: Successfully set pstate[0x54]
00000019.963844421|    1|IMP             |323654000|pgpe_start_suspend: scheduling enable of pstates (owner=0x00)
00000019.964062208|    0|INF             |323700640|task_dcm_icc: Message sequence didn't change
00000019.964117669|    1|IMP             |323710464|SMGR_obs_to_active: Pstates are enabled, continuing with state trans
00000019.964120108|    1|IMP             |323770784|proc_pstate_kvm_setup: populate static OPAL data
00000019.964145617|    1|IMP             |323672240|populate_parameter_table_data: Copying 20 / 20 parameters to PHYP static data shared memory version[0x10]
00000019.964155550|    1|IMP             |323695680|SMGR: Observation to Active Transition Completed. OCC role = 1
00000019.964158042|    1|IMP             |323764192|SMGR: Standby to Active Transition Completed
00000019.964161224|    1|INF             |323735216|check_for_opal_updates: processor throttle status change - 0x00->0xAA
00000019.964163848|    1|INF             |323788256|check_for_opal_updates: OCC state change 0x01->0x03
00000019.964166935|    1|INF             |323819488|check_for_opal_updates: soft min Pcap = 0x01F4->0x01F4 hard min Pcap = 0x0640->0x0640
00000019.964169671|    1|INF             |323669424|check_for_opal_updates: max Pcap = 0x08C0->0x08C0 active Pcap = 0x0000->0x0000
00000019.964172344|    1|INF             |323773776|check_for_opal_updates: Pcap PSR = 100->100 Pcap source = 0->1
00000019.964413267|    0|INF             |323641808|task_dimm_sm: DIMM temp collection is being started (0x80008000)
00000019.964420634|    0|IMP             |323699088|New MBA throttle max|min N values: mba[0x00800020]
00000019.964614272|    0|INF             |323700640|task_dcm_icc: Message sequence didn't change
00000019.965176481|    0|INF             |323700640|task_dcm_icc: Message sequence didn't change
00000019.965266085|    1|INF             |323735216|check_for_opal_updates: processor throttle status change - 0xAA->0x00
00000019.965539447|    0|IMP             |323731392|update_i2c_lock: OCC has acquired lock for I2C engine 3
00000019.965740439|    0|INF             |323700640|task_dcm_icc: Message sequence didn't change
00000019.965764721|    0|IMP             |323806976|INTR: WOF is being enabled...
00000019.966300869|    0|INF             |323700640|task_dcm_icc: Message sequence didn't change
00000019.966332686|    0|IMP             |323636464|calculate_core_leakage: ECO cores[0x30030000]
00000019.966383233|    0|INF             |323621920|amec_set_freq_range: Mode[0x00] Fmin[2000/0x78] Fmax[3900/0x06]
00000019.966385658|    0|INF             |323746896|amec_set_freq_range:   Fmin@MaxThrot[932/0xB8]
00000019.969474001|    0|INF             |323707264|task_memory_data: Memory temp collection is being started
00000019.970251815|    0|IMP             |323733856|amec_pcap_calc: Node pcap set to 700 watts.
00000019.970254926|    0|INF             |323619824|amec_ppb_fmax_calc num chips changed from[0] to[4] NODE_MHZ_PER_WATT[1]
00000020.000037728|    0|INF             |323740048|amec_perfcount_ocmb_getmc: Missing data.  No support for OCMB[0] utilization.  OCMB freq[0] burst length[0]
00000020.000748819|    1|INF             |323707680|check_for_opal_updates: WOF state change 0x00->0x01
00000020.000752030|    1|INF             |323819488|check_for_opal_updates: soft min Pcap = 0x01F4->0x01F4 hard min Pcap = 0x0640->0x0640
00000020.000754542|    1|INF             |323669424|check_for_opal_updates: max Pcap = 0x08C0->0x08C0 active Pcap = 0x0000->0x02BC
00000020.000756974|    1|INF             |323773776|check_for_opal_updates: Pcap PSR = 100->100 Pcap source = 1->1
00000020.019171166|    0|IMP             |323801504|main_mem_sensors_init: Initialization complete for all 415 valid sensors
00000020.062526842|    1|INF             |323787968|CMDH: cmdh_tmgt_setmodestate: changed state from 1 to 3, mode from 0 to 0
00000020.632012734|    1|INF             |323699392|User selected power limit = 700 set from source 1
00000020.632865024|    0|IMP             |323629008|amec_data_write: PCAP Config data: pcap[700]: data_mask[efc]
00000020.660797728|    1|ERR             |323697520|E>cmdh_send_ambient_temp: ambient read failure status!
00000020.660800789|    1|INF             |323727904|cmdh_send_ambient_temp new altitude 228m reference 1000m WOF ambient_adj_for_altitude[-2]
00000069.790378844|    0|INF             |323751200|notify_host: notification of reason 0x04 has been sent
00000149.519231968|    0|IMP             |323781424|First time throttling for reason[0x00004000] History[0x00004000] freq = 3867
00000260.172393214|    1|INF             |323813968|notify_host: OCCMISC/core_ext_intr not clear yet (register=0x80010000, reason=0x01)
00000266.277623189|    1|INF             |323813968|notify_host: OCCMISC/core_ext_intr not clear yet (register=0x80010000, reason=0x01)
00000445.574939413|    0|ERR             |323758416|E>task_core_data_control: pstate[0x79] update IPC task did not complete successfully, idle?[0] rc[00000001]
00000445.574944960|    0|INF             |323751040|createPgpeErrl: Creating error log in slot[0] max size[3616]
00000445.574947770|    0|IMP             |323703232|createPgpeErrl: OISR0=0x00000020, OISR1=0x1000E550, OCCS2=0x02AE02BC
00000445.574950855|    0|INF             |323817456|addPgpeDataToErrl: clipping pgpe trace size from 1080 to 992
00000445.574980316|    0|IMP             |323646656|Activating reset required state.
00000445.575024398|    0|INF             |323766768|Reporting error @ 0xFFFAB050 with size 2752
00000445.575027210|    0|INF             |323781072|ModID: 0x00000406, RC: 0x000000D3, UserData1: 0x00000001, UserData2: 0x00000000
00000445.575110392|    1|IMP             |323704368|SMGR: All States to Safe Transition Started
00000445.575113432|    1|INF             |323736704|INTR: WOF disabled due to non-error condition. wof_disabled = 0x00000100
00000445.575115962|    1|ERR             |323631472|E>INTR: WOF is being disabled. Reasoncode: 0x00000100
00000445.575118734|    1|INF             |323771392|INTR: WOF disabled, setting frequency ranges
00000445.575122199|    1|INF             |323621920|amec_set_freq_range: Mode[0x00] Fmin[2000/0x78] Fmax[2850/0x45]
00000445.575125976|    1|INF             |323746896|amec_set_freq_range:   Fmin@MaxThrot[932/0xB8]
00000445.590139164|    1|INF             |323704864|disable_all_dimms: DIMM temp collection is being stopped
00000445.590141836|    1|INF             |323802016|occ_i2c_lock_release(engine 3) called
00000445.590144679|    1|INF             |323751200|notify_host: notification of reason 0x02 has been sent
00000445.590147159|    1|IMP             |323701296|update_i2c_lock: OCC has released lock for I2C engine 3
00000445.590149605|    1|IMP             |323632256|SMGR: All States to Safe Transition Completed
00000445.590152837|    1|INF             |323765536|check_for_opal_updates: safe state processor throttle status change - 0x01->0x05
00000445.590155607|    1|INF             |323788256|check_for_opal_updates: OCC state change 0x03->0x04
00000445.590158167|    1|INF             |323791600|check_for_opal_updates: Unable to enable WOF due to WOF disable[0x00000100]
00000450.000465591|    1|INF             |323644384|check_for_opal_updates: Mode change 0x04->0x00
00000465.359252391|    1|INF             |323724592|deleting error @0xFFFAB050 at slot [0]
00000470.000465891|    1|INF             |323644384|check_for_opal_updates: Mode change 0x04->0x00
00000490.000466110|    1|INF             |323644384|check_for_opal_updates: Mode change 0x04->0x00
fsp-trace.c main: failed to parse all files (skipped 2 files)
```

**Comment by prasrang** [comment]
Please copy the complete occ errorlog , and if possible attach all PEL logs

**Comment by lili-lilili** [comment]
> Please copy the complete occ errorlog , and if possible attach all PEL logs

There is none PEL logs when occ entry safe mode.
The previous log contains all the occ traces that I can parse out.
The "occ errorlog" is occ trace? 
If not, can you provide a method of collection?

**Comment by prasrang** [comment]
If you are using BMC, then peltool -l will give the log lists
and peltool -i <id> will dump the log, 
I want the output of complete log dump (peltool -i < id>)
In occ log, there will be pgpe trace hex data, so that I can debug more on pgpe failure.

**Comment by prasrang** [comment]
if you are using fsp , then errl -l (list of errorlog)
errl -d < id> (dump the errorlog

**Comment by prasrang** [comment]
Also can I know steps  of changing the powercap value

**Comment by lili-lilili** [comment]
Set powercap value either before or after host startup can cuase the problem.

1st case:
1. Set powercap by BMC
2. Poweron machine --> boot to OS --> start htx test
3. OCC entry safe mode or cpu checkstop

2st case:
1. Poweron machine --> boot to OS --> start htx test
2. Set powercap by BMC
3. OCC entry safe mode or cpu checkstop

**Comment by lili-lilili** [comment]
Here is test log:

test A:
https://github.com/lili-lilili/share_files/blob/main/240222-pcap/test1-occtool-trace.log
https://github.com/lili-lilili/share_files/blob/main/240222-pcap/test1-peltool.log

testB:
https://github.com/lili-lilili/share_files/blob/main/240222-pcap/test2-occtool_log.log
No pellog is created while testB

**Comment by lili-lilili** [comment]
> Set powercap value either before or after host startup can cuase the problem.
> 
> 1st case:
> 
> 1. Set powercap by BMC
> 2. Poweron machine --> boot to OS --> start htx test
> 3. OCC entry safe mode or cpu checkstop
> 
> 2st case:
> 
> 1. Poweron machine --> boot to OS --> start htx test
> 2. Set powercap by BMC
> 3. OCC entry safe mode or cpu checkstop

BMC set powercap by PLDM

**Comment by prasrang** [comment]
Now I see PGPE error logs. And it is same issue that occ was sending pstate request(higher) greater than psafe value, which is invalid.  So we have  the fix and got merged in 1060. 

I bet the above run doesn't have latest master-p10(1060) hcode-1060. 
Pease try with latest 1060 hcode.

**Comment by lili-lilili** [comment]
> Now I see PGPE error logs. And it is same issue that occ was sending pstate request greater than psafe value, which is invalid. So we have the fix and got merged in 1060.
> 
> I bet the above run doesn't have latest master-p10(1060) hcode-1060. Pease try with latest 1060 hcode.

Yes, hcode is fw1030 now.
I test master-p10 a few days ago and occ enter safe mode too.
I will test mater-p10  and collect the error log again.

**Comment by lili-lilili** [comment]
> Now I see PGPE error logs. And it is same issue that occ was sending pstate request greater than psafe value, which is invalid. So we have the fix and got merged in 1060.
> 
> I bet the above run doesn't have latest master-p10(1060) hcode-1060. Pease try with latest 1060 hcode.

Can you share something on how to analyze the hcode log?

**Comment by prasrang** [comment]
In our HCODE errorlog, will have the below section, that talks about some details of the hcode error.

here in the reference code, if we take the value of last 4 nibble 2E10: 2E => component id => PGPE, 10=> PGPE_ERR_REASON_CODE_VOLTAGE    =     0x10

Then if we take last word(hex 9) value 1013 ==>  PGPE_ERR_EXT_CODE_AVSBUS_VOLTAGE_OUT_OF_BOUNDS                  =   0x1013 .... This tells  voltage value that we are trying to write is out of bound.

"Reference Code":           "BC102E10",
    "Hex Word 2":               "000000E0",
    "Hex Word 3":               "00000100",
    "Hex Word 4":               "00000000",
    "Hex Word 5":               "00200000",
    "Hex Word 6":               "00000000",
    "Hex Word 7":               "00000000",
    "Hex Word 8":               "00000000",
    "Hex Word 9":               "00011013",

**Comment by lili-lilili** [comment]
Here is log for hcode-master:

occtool -p
https://github.com/lili-lilili/share_files/blob/main/240227-pcap/occtool-poll.log

occtool -trace:
https://github.com/lili-lilili/share_files/blob/main/240227-pcap/occtool-trace.log

Only one PEL log is created.
The content is the same as this one: https://github.com/open-power/hcode/issues/5#issuecomment-1922575362

**Comment by prasrang** [comment]
I don't see any pgpe related traces/data. And also with the latest hcode , we are not seeing any out of bound voltage issue.  
I think OCC team should takea look may be Chris can hehlp (@cjcain )

**Comment by cjcain** [comment]
Because a power cap was set, the OCC needs to set a pstate to get the power under the power cap. OCC was asking for a pstate change and PGPE is not returning.

00000544.960344408|    0|ERR | 195|E>task_core_data_control: pstate[0x79] update IPC task did not complete successfully, idle?[0] rc[00000001]
00000544.960347856|    0|INF | 456|createPgpeErrl: Creating error log in slot[0] max size[3616]
00000544.960352810|    0|IMP | 466|createPgpeErrl: OISR0=0x00000020, OISR1=0x1000E550, OCCS2=0x02AE02BD
00000544.960356099|    0|INF | 788|addPgpeDataToErrl: clipping pgpe trace size from 2104 to 992

OCC collects the PGPE trace and attaches it to the log. There are also other PGPE logs that should have your traces @prasrang - I see PGPE trace block in all of these PGPE errors: https://github.com/lili-lilili/share_files/blob/main/240222-pcap/test1-peltool.log

To show all PELs/Error logs (including hidden and informational errors) use: `peltool -lhf`

**Comment by prasrang** [comment]
@cjcain  Above pgpe error was known issue in 1050 driver . I have looked at that  and already commented in https://github.com/open-power/hcode/issues/5#issuecomment-1958997483 . 
The fix is in 1060 driver. 

They tested with 1060 hcode itseems and now they don't see pgpe log , but still occ is going to safe mode. Can you please look at this data.
https://github.com/open-power/hcode/issues/5#issuecomment-1965687635

**Comment by cjcain** [comment]
OCC trace is almost identical: (from https://github.com/open-power/hcode/issues/5#issuecomment-1965687635)

00003336.093914588|    0|ERR | 195|E>task_core_data_control: pstate[0x7E] update IPC task did not complete successfully, idle?[0] rc[00000001]
00003336.093920309|    0|INF | 456|createPgpeErrl: Creating error log in slot[0] max size[3616]
00003336.093923274|    0|IMP | 466|createPgpeErrl: OISR0=0x00000020, OISR1=0x1000E550, OCCS2=0x02C702BC
00003336.093926664|    0|INF | 788|addPgpeDataToErrl: clipping pgpe trace size from 1080 to 992
00003336.093958353|    0|IMP |  96|Activating reset required state.

I suspect the logs are hidden/informational.  Or maybe system crashes before logs got collected?
@lili-lilili need to display all PELs:
To show all PELs/Error logs (including hidden and informational errors) use: peltool -lhf

**Comment by lili-lilili** [comment]
Here is the log:
https://github.com/lili-lilili/share_files/blob/main/240301-pcap/occtool.txt
https://github.com/lili-lilili/share_files/blob/main/240301-pcap/peltool.txt

**Comment by prasrang** [comment]
I don't see any pgpe logs/trace.

can you dump the output of  "cat /media/hostfw/running/VERSION"

**Comment by cjcain** [comment]
Again the occ traces from occtool.txt do show same issue (as expected).  OCC was able to set pstate at timestamp 84, but failed at 1906:
00000084.845922007|    0|IMP | 911|First time throttling for reason[0x00004000] History[0x00004000] freq = 3884
00001906.658556792|    0|ERR | 195|E>task_core_data_control: pstate[0x7B] update IPC task did not complete successfully, idle?[0] rc[00000001]
00001906.658560481|    0|INF | 456|createPgpeErrl: Creating error log in slot[0] max size[3616]
00001906.658565690|    0|IMP | 466|createPgpeErrl: OISR0=0x00000020, OISR1=0x1000E550, OCCS2=0x02C702BC
00001906.658568858|    0|INF | 788|addPgpeDataToErrl: clipping pgpe trace size from 1080 to 992

There are no OCC or PGPE in the collected peltool.txt file which is not normal. I suspect there is a communication issue with Hostboot (BMC not able to retrieve any logs from Hostboot).

Looking at the BD122682 logs (from BMC) I see:
        "Mar 01 15:13:55 openpower-occ-control[653]: safeStateDelayExpired: OCC2 is in SAFE state, requesting reset",
        "Mar 01 15:13:55 openpower-occ-control[653]: safeStateDelayExpired: OCC3 is in SAFE state, requesting reset",
        "Mar 01 15:13:55 openpower-occ-control[653]: sendPldm: calling pldm_send(mctpID:10, fd:20, 8 bytes) for OCC2",
        "Mar 01 15:13:55 openpower-occ-control[653]: Status::occActive OCC2 changed to false",
        "Mar 01 15:13:55 openpower-occ-control[653]: >>Status::resetOCC() - requesting reset for OCC2",
 ...
        "Mar 01 15:13:56 openpower-occ-control[653]: pldm: got new InstanceId: 1",
        "Mar 01 15:13:56 openpower-occ-control[653]: sendPldm: calling pldm_send(mctpID:10, fd:20, 8 bytes) for OCC3",
        "Mar 01 15:13:56 openpower-occ-control[653]: Status::occActive OCC3 changed to false",
        "Mar 01 15:13:56 openpower-occ-control[653]: >>Status::resetOCC() - requesting reset for OCC3",
        "Mar 01 15:13:56 phosphor-log-manager[538]: Bad PLDM completion code",
        "Mar 01 15:13:56 phosphor-log-manager[538]: Event not found in PEL message registry: xyz.openbmc_project.Logging.SEL.Error.Created",
        "Mar 01 15:13:56 phosphor-log-manager[538]: PLDM command response failure",

A request was made to reset the PM Complex (OCCs), but that message must not have made it to Hostboot (or it wasnt running).  This would also indicate some issue with communication with Hostboot.

@lili-lilili Are you able to collect a hostboot runtime trace? Do you have cronus (to read debug data)?

@prasrang Is there a way he could collect the PGPE trace from the failed state?

**Comment by lili-lilili** [comment]
> I don't see any pgpe logs/trace.
> 
> can you dump the output of "cat /media/hostfw/running/VERSION"

Here is version info:
https://github.com/lili-lilili/share_files/blob/main/240301-pcap/version.log

Most of the module version come from fw1030, occ version is fw1050, hcode version is master-p10

**Comment by lili-lilili** [comment]
> @lili-lilili Are you able to collect a hostboot runtime trace? Do you have cronus (to read debug data)?

I have cronus.
Can you provide cmd to "read debug data"?

Does “Runtime trace“  refer to "opal-prd log"?

**Comment by prasrang** [comment]
If you have cronus setup then it's good. can you get me the pgpe sram dump

From cronus please run this

getsram fff20000 65536 -ch 0 -p0 -fb pgpe_sram.bin

Please attach the pgpe_sram.bin

**Comment by cjcain** [comment]
The HBRT trace would be in the opal-prd log files. (could be var/log/opal-prd, /var/log/syslog or /var/log/messages)
This command may show which file: sudo service opal-prd status
I believe you would see "HTMGT" traces in the appropriate file.

**Comment by lili-lilili** [comment]
> If you have cronus setup then it's good. can you get me the pgpe sram dump
> 
> From cronus please run this
> 
> getsram fff20000 65536 -ch 0 -p0 -fb pgpe_sram.bin
> 
> Please attach the pgpe_sram.bin

Here is pgpe dump:
https://github.com/lili-lilili/share_files/blob/main/240306-pcap/pgpe_sram.bin

**Comment by lili-lilili** [comment]
> The HBRT trace would be in the opal-prd log files. (could be var/log/opal-prd, /var/log/syslog or /var/log/messages) This command may show which file: sudo service opal-prd status I believe you would see "HTMGT" traces in the appropriate file.

Unfortunately, no opal prd log were found.
1. boot to OS
2. rm all the prd log
3. start test, and occ entry safe mode
4. check opal-prd log, i can only find an empty dir for opal-prd log

**Comment by cjcain** [comment]
Did you confirm that the opal-prd service was running before starting the test?  If so, there should have been traces. (check status with: sudo service opal-prd status)
If it is running, send the full status output from as well as the 3 files: /var/log/opal-prd, /var/log/syslog or /var/log/messages

**Comment by lili-lilili** [comment]
> Did you confirm that the opal-prd service was running before starting the test? If so, there should have been traces. (check status with: sudo service opal-prd status) If it is running, send the full status output from as well as the 3 files: /var/log/opal-prd, /var/log/syslog or /var/log/messages

I  confirm opal-prd is running.
If opal-prd doesn't work, the occtool will not work too.

**Comment by lili-lilili** [comment]
And during OS boot, OPAL prd log will be generated.
But no opal-prd log is print when htx test going.

**Comment by prasrang** [comment]
@cjcain  Is it possible to tell which OCC failed ( root cause for safe mode) ?

**Comment by cjcain** [comment]
Safe mode usually happens because multiple resets fail to recover from failure.
Since you only dumped OCC0 trace, that does indicate that the OCC0 (first proc) encountered the error.

From the 2682 PEL:
        "Mar 01 14:41:25 openpower-occ-control[653]: validateOccMaster: OCC0 is master of 4 OCCs",
        "Mar 01 14:41:30 openpower-occ-control[653]: Status::readOccState: OCC1 state 0x03 (lastState: 0x00)",
        "Mar 01 14:41:30 openpower-occ-control[653]: Status::readOccState: OCC2 state 0x03 (lastState: 0x00)",
        "Mar 01 14:41:30 openpower-occ-control[653]: Status::readOccState: OCC3 state 0x03 (lastState: 0x00)",
        "Mar 01 15:12:55 openpower-occ-control[653]: Status::readOccState: OCC2 state 0x04 (lastState: 0x03)",
        "Mar 01 15:12:55 openpower-occ-control[653]: Status::readOccState: OCC3 state 0x04 (lastState: 0x03)",
        "Mar 01 15:13:00 openpower-occ-control[653]: Status::readOccState: OCC0 state 0x04 (lastState: 0x03)",
        "Mar 01 15:13:00 openpower-occ-control[653]: Status::readOccState: OCC1 state 0x04 (lastState: 0x03)",
        "Mar 01 15:13:55 openpower-occ-control[653]: safeStateDelayExpired: OCC2 is in SAFE state, requesting reset",
That shows that at 15:12:55 (31 minutes after going active), all OCCs changed to SAFE state.

This does not necessarily say that OCC2 hit it first, but just that the BMC saw the OCC2 state change first.
Since OCC0 is the master, a failure on that OCC could trigger failures on the slaves as well.
You could dump all the OCC traces for review

Please send:
1. output of this command: sudo service opal-prd status
2. /var/log/syslog
3. /var/log/messages
4. /var/log/opal-prd (unless completely empty)

**Comment by lili-lilili** [comment]
> Safe mode usually happens because multiple resets fail to recover from failure. Since you only dumped OCC0 trace, that does indicate that the OCC0 (first proc) encountered the error.
> 
> From the 2682 PEL: "Mar 01 14:41:25 openpower-occ-control[653]: validateOccMaster: OCC0 is master of 4 OCCs", "Mar 01 14:41:30 openpower-occ-control[653]: Status::readOccState: OCC1 state 0x03 (lastState: 0x00)", "Mar 01 14:41:30 openpower-occ-control[653]: Status::readOccState: OCC2 state 0x03 (lastState: 0x00)", "Mar 01 14:41:30 openpower-occ-control[653]: Status::readOccState: OCC3 state 0x03 (lastState: 0x00)", "Mar 01 15:12:55 openpower-occ-control[653]: Status::readOccState: OCC2 state 0x04 (lastState: 0x03)", "Mar 01 15:12:55 openpower-occ-control[653]: Status::readOccState: OCC3 state 0x04 (lastState: 0x03)", "Mar 01 15:13:00 openpower-occ-control[653]: Status::readOccState: OCC0 state 0x04 (lastState: 0x03)", "Mar 01 15:13:00 openpower-occ-control[653]: Status::readOccState: OCC1 state 0x04 (lastState: 0x03)", "Mar 01 15:13:55 openpower-occ-control[653]: safeStateDelayExpired: OCC2 is in SAFE state, requesting reset", That shows that at 15:12:55 (31 minutes after going active), all OCCs changed to SAFE state.
> 
> This does not necessarily say that OCC2 hit it first, but just that the BMC saw the OCC2 state change first. Since OCC0 is the master, a failure on that OCC could trigger failures on the slaves as well. You could dump all the OCC traces for review
> 
> Please send:
> 
> 1. output of this command: sudo service opal-prd status
> 2. /var/log/syslog
> 3. /var/log/messages
> 4. /var/log/opal-prd (unless completely empty)

All of the logs were collected:
https://github.com/lili-lilili/share_files/tree/main/240312-pcap

**Comment by lili-lilili** [comment]
> @cjcain Is it possible to tell which OCC failed ( root cause for safe mode) ?

Can you find anything in the sram?
Does the sram have a pgpace trace?

**Comment by cjcain** [comment]
Pstates/frequencies from the PGPE headers:
00000000.000360078|    1|IMP |1111|MAIN: read_oppb_params: PGPE Frequency data: min[1999000], max[4000000], step[16667] kHz, pState min[0x78]
00000000.000362163|    1|IMP | 194|MAIN: populate_sys_mode_freq_table: Clipping freq to system max ceiling[3900000KHz] < Pstate0[4000000KHz]
00000000.000364524|    1|IMP | 257|MAIN: populate_sys_mode_freq_table: clipping chip Fmax[4000MHz] to system max[3900MHz]
00000000.000366487|    1|ERR | 269|E>MAIN: populate_sys_mode_freq_table: chip min[2000MHz] is higher than sys min[1999MHz]
00000000.000368848|    1|IMP | 384|MAIN: populate_sys_mode_freq_table: freq points min[2000MHz], WOF Base[2850MHz], UT[3900MHz], max[3900MHz]
00000000.000370960|    1|IMP | 396|MAIN: populate_sys_mode_freq_table: Mode freq disabled[2600MHz], Psav[2000MHz], Perf[3900MHz], Fmax[3900MHz]
00000000.000378065|    1|IMP |1125|MAIN: read_oppb_params:  pstate_max_throttle[0xB8]/1999000kHz(64 steps)  Fmin Pstate[0x78]

All 4 OCCs are hitting the same failure trying to go to pstate 0x79 (timestamps are not synced between occs, so could be off)
occ0:
00000811.848912784|    0|ERR | 195|E>task_core_data_control: pstate[0x79] update IPC task did not complete successfully, idle?[0] rc[00000001]
occ1:
00000811.848912784|    0|ERR | 195|E>task_core_data_control: pstate[0x79] update IPC task did not complete successfully, idle?[0] rc[00000001]
occ2:
00000810.313609333|    0|ERR | 195|E>task_core_data_control: pstate[0x79] update IPC task did not complete successfully, idle?[0] rc[00000001]
occ3:
00000804.693532480|    0|ERR | 195|E>task_core_data_control: pstate[0x79] update IPC task did not complete successfully, idle?[0] rc[00000001]

Pstate 0x79 is valid, but it would be the first pstate in the throttle range.  (since 0x78 is the min frequency)

@prasrang Do you see anything in the pgpe traces?

**Comment by prasrang** [comment]
I think , there might be issue in pgpe in handling throttle range, when WOF is disabled dynamically and new clip update request has come. Still debugging.

**Comment by prasrang** [comment]
I think  I have the fix. There was a bug in handling throttle clip bound. I have pushed the fix to our HCODE repository. THis might  take some time to merge into open-power repo.

@lili-lilili Will you able to compile hcode and patch to your system. 
The fix is in the file "chips/p10/procedure/ppe/pgpe/pgpe_pstate.c"
Function :pgpe_pstate_is_clip_bounded 
If you can replace the below function in pgpe_pstate.c and build and test.

uint32_t pgpe_pstate_is_clip_bounded()
{
    uint32_t clip_min = G_pgpe_pstate.clip_min > G_pgpe_pstate.pstate_safe ? G_pgpe_pstate.pstate_safe :
                        G_pgpe_pstate.clip_min;

    uint32_t ps_bounded = 0;
    uint32_t thr_bounded = 1;

    if ((G_pgpe_pstate.pstate_curr >= clip_min) &&
        (G_pgpe_pstate.pstate_curr <= G_pgpe_pstate.clip_max))
    {
        ps_bounded = 1;
    }

    // If throttle is pending , then only we need to check throttle bounded or
    // not.
    if (pgpe_thr_ctrl_is_enabled() &&
        G_pgpe_pstate.throttle_pending) // this is the fix
    {
        if(G_pgpe_pstate.throttle_clip > G_pgpe_pstate.throttle_curr)
        {
            thr_bounded = 0;
        }
    }

    return (thr_bounded && ps_bounded);
}

**Comment by lili-lilili** [comment]
That's really good news.
I will try your fix and give feedback on the test results.

**Comment by lili-lilili** [comment]
@prasrang It looks effective.
The test has been running for several hours and the occ is still active.
I will go on test for more time, hope it work will.

Ask a few questions:
1. How to debug PGPE SRAM?
If we can get some trace or info from sram?

2. Can you point out the commit to fix this issue?
https://github.com/open-power/hcode/issues/5#issuecomment-1926287985
I want to fix the issude at hcode fw1030 (the version we used now).

**Comment by prasrang** [comment]
This is the commit I have pushed to gerrit , not sure you will have access
https://rchgit01.rchland.ibm.com/gerrit1/c/hw/ekb/+/167657

**Comment by lili-lilili** [comment]
> This is the commit I have pushed to gerrit , not sure you will have access https://rchgit01.rchland.ibm.com/gerrit1/c/hw/ekb/+/167657

Looks like i don't have permission.
I remember you said that the fix has been merged into master-p10.
Can you provide the commit on the github?

**Comment by lili-lilili** [comment]
> @cjcain Above pgpe error was known issue in 1050 driver . I have looked at that and already commented in [#5 (comment)](https://github.com/open-power/hcode/issues/5#issuecomment-1958997483) . The fix is in 1060 driver.
> 
> They tested with 1060 hcode itseems and now they don't see pgpe log , but still occ is going to safe mode. Can you please look at this data. [#5 (comment)](https://github.com/open-power/hcode/issues/5#issuecomment-1965687635)

I mean, this fix.

**Comment by lili-lilili** [comment]
@prasrang

**Comment by prasrang** [comment]
Sorry I didn't get your question. Are you asking is fix will be available in 1050 driver ?

**Comment by lili-lilili** [comment]
> Sorry I didn't get your question. Are you asking is fix will be available in 1050 driver ?

Now we know two issues related to powercapping. 

The second one: 
https://github.com/open-power/hcode/issues/5#issuecomment-2002060916

The first one: 
https://github.com/open-power/hcode/issues/5#issuecomment-1925353331
I would like to know which commit fixed this issue.

**Comment by prasrang** [comment]
Both the fixes are related and the latest fix I gave was the extension. 

Latest fix is also merged in our ekb rebo. Hopefully it should be in openpower repo in another week I think.

**Comment by lili-lilili** [comment]
You mean this commit?
https://github.com/open-power/hcode/commit/adf20b173b9250b63349e6554174b3a0184bb8fd

It  has been merged to master-p10.
If i patch it to fw1030.
The occ will still entry safe mode.
Do I need to merge other patches?

**Comment by prasrang** [comment]
Correct the above commit is merged.
[pgpe_pstate.c](https://github.com/open-power/hcode/files/14745008/pgpe_pstate.c)

I doubt master-p10 code works on fw1030. Because master-p10 will have more changes compared to 1030.

One thing we can do , I have picked 1030 hcode file pgpe_pstate.c. if you can pick this and build and try

**Comment by prasrang** [comment]
Correct the above commit is merged
[pgpe_pstate.txt](https://github.com/open-power/hcode/files/14745025/pgpe_pstate.txt)


I doubt master-p10 code works on fw1030. Because master-p10 will have more changes compared to 1030.

One thing we can do , I have picked 1030 hcode file pgpe_pstate.c. if you can pick this and build and try.
As I couldn't upload in .c format, hence  I have renamed as .txt. After you download you can rename to .c and compare the diff with your code base and apply the changes.

**Comment by lili-lilili** [comment]
It works.
Thanks for your powerful help.
